docs/design/design.md

```diff
@@ -0,0 +1,22 @@
+# Design Documentation
+
+This is an attempt to document the OpenShift Installer's existing design principles and also articulate design goals to guide with new feature development and in some cases inform bug fixes.
+
+## Design Goals
+
+### Install methods
+The OpenShift Installer currently supports 2 installation types, UPI (User Provisioned Infrastructure) and IPI (Installer Provisionined Infrastructure). IPI is a highly opinionated method of installing OpenShift. The IPI installation method tries to automate a large portion of the OpenShift Install with the least amount of intervention from the user. While it tries to take care of supporting the most common install scenarios, it cannot be expected to support every install variation that a user wants to use for their cloud.
+
+This is where the UPI offers the greater flexibility required for these highly specialized installs. In this install method, the user is expected to provide and manage several non-OpenShift components needed for the install on their own. For this reason, documentation surrounding the UPI flow are especially important.
+
+The Openshift Installer has the responsibility to validate configuration provided to it. This validation step is common to both the UPI and IPI install methods.
+
+### Cluster Configuration
+
+It is a design goal for the Installer to obtain the simplest and most minimal configuration required to fully create an OpenShift cluster on the platform and with the capabilities requested by the user. So, for that reason, any information that can be learned from the install environment should not be requested as an input configuration during the install.
```

Comment:
I like that we are stating a design goal, but I have two points here: 
1. I think it would be helpful to establish the motivation and guiding principles for the goal; which in the case of simplicity  and especially for discovering configuration, the principle would be something along the lines of single-responsibility; if we can defer the responsibility to another entity, we opt for that rather than having the installer become a monolith
2. I don't think "simple and minimal" is a good goal because it doesn't provide inherent value; "user-friendly" may be a suitable alternative

Reply:
Agreed, "user-friendly" is not great. I think the current wording:

> most minimal configuration required to fully create an OpenShift cluster on the platform and with the capabilities requested by the user 

is pretty good. I _think_ the objective for this document is so that someone can use it as a guide when answering the question: "should this knob go in the install config?" If we say the guiding principal is to keep the install config _minimal_, then the answer would always be no... in fact we should be removing fields!

The current wording is better in that it provides a little more wiggle room, but I do think we can do better, although I think it will be hard... (and also, again, assuming I am on the right track for what we are trying to accomplish).  

Another issue I have with minimal is that I actually don't think that should be a long-term goal. I think the idea that config should be "minimal" has been a defensive maneuver. to allow us to limit scope/focus; we have said we will limit our focus to a subset of features, power users can still go in and edit manifests, but there are no validations/guard rails in that case. But if we had the ability, I don't see why we wouldn't allow more appropriate config, assuming that it belong in the install config in the first place and we were able to validate the inputs. I think the area this becomes most apparent is for the machineset manifests: the machine API provides a set of features for configuring machines, and we mirror a subset of those in the install config. For example, in [AWS security groups](https://github.com/openshift/api/blob/master/machine/v1beta1/types_awsprovider.go#L60) are configurable. If we were able to safely allow that configuration, why wouldn't we include it in the install config? In fact, I think a fair amount of our work for the foreseeable future will be to expand our config in exactly this way. Ideally, IMHO, we would find a way to unify the efforts between the machine api operator and installer so that API definitions and validations are written once and then vendored in, but I get the sense that is a tall order.

Here are some starter criteria for things I think should belong in the install config, hopefully we will be able to think of more:
1. User provided - if the config can be inferred, it should be; also, general cluster config is generated by the [cluster config operator](https://github.com/openshift/cluster-config-operator)  on the bootstrap node, this has come up in PRs like https://github.com/openshift/installer/pull/5676
2. Input should be verifiable - we should be able to run validation to ensure that config, especially as a whole, will result in a valid cluster